### PR TITLE
Refactor Ky, rebase the `retry` option

### LINKED ISSUE #2: More comprehensive `retry` option
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/ky/issues/2)

In addition to accepting a number, it should accept an object with the ability to specify which methods and status codes to retry on. Same as Got: https://github.com/sindresorhus/got#retry

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/ky/issues/2).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/ky/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by szmarczak**

### CODE REVIEW

**Comment by szmarczak** [comment]
At this point I cannot continue because of the other PRs. We need to merge them first, so there will be less conflict.

**Comment by sindresorhus** [comment]
Any chance you could do the refactoring in a separate PR from the `retry` option improvements? That would make it easier to review.

**Comment by sindresorhus** [comment]
@szmarczak Still interested in doing this refactor? Note that #110 landed which conflicts with this.

**Comment by szmarczak** [comment]
@sindresorhus Sure, I'll do this refactor ASAP.

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
 	"scripts": {
 		"build": "rollup index.js --format=umd --name=ky --file=umd.js",
 		"prepare": "npm run build",
-		"test": "xo && npm run build && nyc ava && tsd"
+		"test": "xo && npm run build && tsd && nyc ava"
```
Comment (szmarczak) [inline package.json]: This is to allow: `npm test test/main.js`

File: readme.md [hunk 2]
```diff
@@ -174,22 +174,23 @@ import ky from 'ky';
...
-
+Type: `number | object`<br>
 Default:
-- `limit`: `2`
-- `methods`: `get` `put` `head` `delete` `options` `trace`
-- `statusCodes`: [`408`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408) [`413`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/413) [`429`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) [`500`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500) [`502`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/502) [`503`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503) [`504`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504)
-- `maxRetryAfter`: `undefined`
-
-An object representing `limit`, `methods`, `statusCodes` and `maxRetryAfter` fields for maximum retry count, allowed methods, allowed status codes and maximum [`Retry-After`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) time.
+- retries: `2`
```
Comment (szmarczak) [inline readme.md]: I know there needs to be a `limit` and `retryFunction` property... I'll do a follow-up PR once https://github.com/sindresorhus/got/pull/830 is done

File: test/retry.js [hunk 3]
```diff
@@ -124,7 +131,7 @@ test('respect 413 Retry-After with timestamp', async t => {
 	});
 
 	const result = await ky(server.url).text();
-	t.true(Number(result) >= retryAfterOn413 * 1000);
+	t.true(Number(result) >= (retryAfterOn413 - 1) * 1000);
```
Comment (szmarczak) [inline test/retry.js]: Yes, we need `-1`. To round the result. You can be +500ms into the next second and the result can be `1500`ms, which is still valid.

File: test/retry.js [hunk 4]
```diff
@@ -158,7 +165,7 @@ test('respect number of retries', async t => {
 
 	await t.throwsAsync(ky(server.url, {
 		retry: {
-			limit: 3
```
Comment (szmarczak) [inline test/retry.js]: This is a **RETRY** limit, not the request count.

**Comment by szmarczak** [comment]
> At this point I cannot continue because of the other PRs.

Not sure what I wanted to continue with. I wish I had written that down.

**Comment by sholladay** [comment]
This is probably too stale at this point. Thoughts, @szmarczak?

On the upside, I took inspiration from your refactoring for other PRs, so the options processing/validation should be a bit easier to work with now, especially since PR #180 was merged.

**Pull request closed without merging**